router/module.go

```diff
@@ -225,117 +225,162 @@ func (am AppModule) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string
+			"dst-channel", packet.DestinationChannel, "dst-port", packet.DestinationPort,
+			"error", err,
+		)
+		return am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
 	}
 
+	am.keeper.Logger(ctx).Debug("packetForwardMiddleware OnAcknowledgementPacket",
+		"sequence", packet.Sequence,
+		"src-channel", packet.SourceChannel, "src-port", packet.SourcePort,
+		"dst-channel", packet.DestinationChannel, "dst-port", packet.DestinationPort,
+		"amount", data.Amount, "denom", data.Denom,
+	)
+
 	var ack channeltypes.Acknowledgement
-	if err := json.Unmarshal(acknowledgement, &ack); err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrJSONUnmarshal, err.Error())
+	if err := channeltypes.SubModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet acknowledgement: %v", err)
 	}
 
-	if !ack.Success() {
-		// If acknowledgement indicates error, no retries should be attempted. Refund will be initiated now.
-		am.keeper.RefundForwardedPacket(ctx, packet, am.refundTimeout)
-		return nil
+	inFlightPacket := am.keeper.GetAndClearInFlightPacket(ctx, packet.SourceChannel, packet.SourcePort, packet.Sequence)
+	if inFlightPacket != nil {
+		// this is a forwarded packet, so override handling to avoid refund from being processed.
+		return am.keeper.WriteAcknowledgementForForwardedPacket(ctx, packet, data, inFlightPacket, ack)
 	}
 
-	// For successful ack, we no longer need to track this packet.
-	am.keeper.RemoveInFlightPacket(ctx, packet)
-	return nil
+	return am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
 }
 
 // OnTimeoutPacket implements the IBCModule interface
 func (am AppModule) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) error {
-	if err := am.app.OnTimeoutPacket(ctx, packet, relayer); err != nil {
-		return err
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		am.keeper.Logger(ctx).Error("packetForwardMiddleware error parsing packet data from timeout packet",
+			"sequence", packet.Sequence,
+			"src-channel", packet.SourceChannel, "src-port", packet.SourcePort,
+			"dst-channel", packet.DestinationChannel, "dst-port", packet.DestinationPort,
+			"error", err,
+		)
+		return am.app.OnTimeoutPacket(ctx, packet, relayer)
 	}
 
-	if err := am.keeper.HandleTimeout(ctx, packet); err != nil {
-		am.keeper.RefundForwardedPacket(ctx, packet, am.refundTimeout)
+	am.keeper.Logger(ctx).Debug("packetForwardMiddleware OnAcknowledgementPacket",
+		"sequence", packet.Sequence,
+		"src-channel", packet.SourceChannel, "src-port", packet.SourcePort,
+		"dst-channel", packet.DestinationChannel, "dst-port", packet.DestinationPort,
+		"amount", data.Amount, "denom", data.Denom,
+	)
+
+	inFlightPacket, err := am.keeper.TimeoutShouldRetry(ctx, packet)
```

Comment:
Deferring to the intended users here.

But it seems to me that this may have been requested before async acks were used? Now with async acks, multihop is atomic. It either reaches the final destination or it ends back at sender chain. So i don't see a need to implement retries directly in-protocol?

I think its cleaner to have this base layer in the state machine just retry once, and if it fails, then the retry logic can be implemented by callers on the sender chain

Reply:
The retry on timeout logic is more useful after point-of-no-return actions, such as swap before a forward. Since we cannot refund back to the sender chain after a successful swap, it is helpful to retry on timeout for the forwards after the swap.

Reply:
Interesting... it seems to me like doing something complicated in the middle is a bit dangerous with async acks.

Suppose you swap in the middle and then it times out at the destination (let's suppose it does for all of your tries).

If the error ack is sent, then the tokens are refunded in their original denom. Even though the original denom funds on the router chain got put into a DEX and were swapped out for something new. What would happen now to those swapped funds?

The scheme is fine in the case where the chains in the middle are acting as pure pass-throughs. I'm worried about the potential issues that would arise if the chains in the middle are doing something more complicated.

Is this an expected use case people need? Is it fine to just swap at the final destination chain for most use cases?

Reply:
> If the error ack is sent, then the tokens are refunded in their original denom. Even though the original denom funds on the router chain got put into a DEX and were swapped out for something new. What would happen now to those swapped funds?

For the swap and forward case, the atomicity from the sender will only be up to the swap. A successful ack will be proxied back in the case of a successful swap, regardless of the success/fail of the forward after the swap. This will be necessary with or without timeout on retry so that refunds don't happen after a successful swap.

> Is this an expected use case people need? Is it fine to just swap at the final destination chain for most use cases?

The primary use case of swap and forward will be a swap on chain B in an A->B->C transfer. E.g. start with native denom on A, swap ibc denom A for ibc denom C on chain B dex, then forward to C to result in native denom C on chain C. This can all occur from a single user initiated transfer on A.

> I'm unfamiliar with async acks (where can I find information about this?)

There are a few diagrams in the PR description with numbers to indicate the steps of the packet flow, showing the acks being held until the response of the ack of the forwarded packet.

> Is the idea of retries after point-of-no-return just to reduce the likelihood of having to end up in this case (due to congestion or temporary liveness issues, for example)?

The retries can be configured for any of the hops, not just after a point of no return. The chain sets the defaults, they are optional to be included in the packet memo (could be 0 for no retries). But yes exactly, it helps for relayer liveness issues.

Reply:
> A successful ack will be proxied back in the case of a successful swap, regardless of the success/fail of the forward after the swap. This will be necessary with or without timeout on retry so that refunds don't happen after a successful swap.

To clarify this: we can still retain atomic behavior for swap and forward. In the swap middleware, we can hold the ack until the transfer after the swap returns an ack. Then, regardless if there is success/error in the forward(s) after the swap, we send back a successful ack, but it can contain a specific error message for failures after the swap. That way there is no risk of funds being refunded on prior chains after the swap is successful and the user can still poll for the ack on the first chain and check the ack message for success/fail of the whole operation. For example, if the A->B->C example above has a successful swap on B but has a failure on the forward B to C, you would get a successful ack on A with message, "Failed to forward to chain C after max retries after a successful swap on B. Funds are now in denom CB on chain B"